opentelemetry/src/attributes/set.rs

```diff
@@ -0,0 +1,363 @@
+                                .iter()
+                                .map(|x| OrderedFloat(*x))
+                                .collect::<Vec<_>>()
+                                .cmp(&of.iter().map(|x| OrderedFloat(*x)).collect()),
+                            other => other,
+                        }
+                    }
+                    (Value::Bool(b), Value::Bool(ob)) => b.cmp(ob),
+                    (Value::I64(i), Value::I64(oi)) => i.cmp(oi),
+                    (Value::String(s), Value::String(os)) => s.cmp(os),
+                    _ => Ordering::Equal,
+                },
+                other => other, // 2nd order by value types
+            },
+            other => other, // 1st order by key
+        }
+    }
+}
+
+fn type_order(v: &Value) -> u8 {
+    match v {
+        Value::Bool(_) => 1,
+        Value::I64(_) => 2,
+        Value::F64(_) => 3,
+        Value::String(_) => 4,
+        Value::Array(a) => match a {
+            Array::Bool(_) => 5,
+            Array::I64(_) => 6,
+            Array::F64(_) => 7,
+            Array::String(_) => 8,
+        },
+    }
+}
+
+impl PartialEq for HashKeyValue {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.key == other.0.key
+            && match (&self.0.value, &other.0.value) {
+                (Value::F64(f), Value::F64(of)) => OrderedFloat(*f).eq(&OrderedFloat(*of)),
+                (Value::Array(Array::F64(f)), Value::Array(Array::F64(of))) => {
+                    f.len() == of.len()
+                        && f.iter()
+                            .zip(of.iter())
+                            .all(|(f, of)| OrderedFloat(*f).eq(&OrderedFloat(*of)))
+                }
+                (non_float, other_non_float) => non_float.eq(other_non_float),
+            }
+    }
+}
+
+impl Eq for HashKeyValue {}
+
+static EMPTY_SET: Lazy<Arc<InternalAttributeSet>> =
+    Lazy::new(|| Arc::new(InternalAttributeSet::new(Vec::with_capacity(0))));
+
+#[derive(Eq, PartialEq, Debug)]
+struct InternalAttributeSet {
+    key_values: Vec<HashKeyValue>,
+    hash: u64,
+}
+
+impl InternalAttributeSet {
+    fn new(mut values: Vec<HashKeyValue>) -> Self {
+        values.sort_unstable();
+        let mut hasher = DefaultHasher::new();
+        values.iter().fold(&mut hasher, |mut hasher, item| {
+            item.hash(&mut hasher);
+            hasher
+        });
+
+        InternalAttributeSet {
+            key_values: values,
+            hash: hasher.finish(),
+        }
+    }
+}
+
+impl<I: IntoIterator<Item = KeyValue>> From<I> for InternalAttributeSet
+where
+    <I as IntoIterator>::IntoIter: DoubleEndedIterator + ExactSizeIterator,
+{
+    fn from(value: I) -> Self {
+        let iter = value.into_iter();
+        let mut seen_keys = HashSet::with_capacity(iter.len());
+        let vec = iter
+            .into_iter()
+            .rev()
+            .filter_map(|kv| {
+                if seen_keys.insert(kv.key.clone()) {
```

Comment:
Not sure that we need to clone here since we own the reference.

Reply:
fixed